core/src/main/java/org/apache/calcite/rel/metadata/RelMdFunctionalDependency.java

```diff
@@ -64,212 +78,353 @@ protected RelMdFunctionalDependency() {}
     return BuiltInMetadata.FunctionalDependency.DEF;
   }
 
+  /**
+   * Determines if column is functionally dependent on key for a given rel node.
```

Comment:
if this is a conservative approximation, please document if it is an over- or under- approximation, or if it is exact. I think this comments applies to many other methods in this class.